src/amp-story-player.js

```diff
@@ -432,6 +469,187 @@ export class AmpStoryPlayer {
+      if (!this.touchEventState_.isSwipeX) {
+        return;
+      }
+    }
+
+    this.onSwipeX_({
+      deltaX: x - this.touchEventState_.startX,
+      last: false,
+    });
+  }
+
+  /**
+   * Reacts to touchend events. Resets cached touch event states.
+   * @private
+   */
+  onTouchEnd_() {
+    if (this.touchEventState_.isSwipeX === true) {
+      this.onSwipeX_({
+        deltaX: this.touchEventState_.lastX - this.touchEventState_.startX,
+        last: true,
+      });
+    }
+
+    this.touchEventState_.startX = 0;
+    this.touchEventState_.startY = 0;
+    this.touchEventState_.lastX = 0;
+    this.touchEventState_.isSwipeX = null;
+  }
+
+  /**
+   * Reacts to horizontal swipe events.
+   * @param {!Object} gesture
+   */
+  onSwipeX_(gesture) {
+    const {deltaX} = gesture;
+
+    if (gesture.last === true) {
+      const delta = Math.abs(deltaX);
+
+      if (this.swipingState_ === SwipingState.SWIPING_TO_NEXT) {
+        delta > TOGGLE_THRESHOLD_PX && this.getSecondaryIframe_()
+          ? this.next_()
+          : this.resetIframeStyles_();
+      }
+
+      if (this.swipingState_ === SwipingState.SWIPING_TO_PREVIOUS) {
+        delta > TOGGLE_THRESHOLD_PX && this.getSecondaryIframe_()
+          ? this.previous_()
+          : this.resetIframeStyles_();
+      }
+
+      return;
+    }
+
+    this.drag_(deltaX);
+  }
+
+  /**
+   * Resets styles for the currently swiped iframes.
+   * @private
+   */
+  resetIframeStyles_() {
+    const currentIframe = this.iframes_[
+      this.stories_[this.currentIdx_][IFRAME_IDX]
+    ];
+
+    resetStyles(currentIframe, ['transform', 'transition']);
+
+    const secondaryIframe = this.getSecondaryIframe_();
+    if (secondaryIframe) {
+      resetStyles(secondaryIframe, ['transform', 'transition']);
+    }
+  }
+
+  /**
+   * Gets accompanying iframe for the currently swiped iframe if any.
+   * @private
+   * @return {?IframeElement}
+   */
+  getSecondaryIframe_() {
+    const nextStoryIdx =
+      this.swipingState_ === SwipingState.SWIPING_TO_NEXT
+        ? this.currentIdx_ + 1
+        : this.currentIdx_ - 1;
+
+    if (nextStoryIdx < 0 || nextStoryIdx >= this.stories_.length) {
+      return;
+    }
+
+    return this.iframes_[this.stories_[nextStoryIdx][IFRAME_IDX]];
+  }
+
+  /**
+   * Drags stories following the swiping gesture.
+   * @param {number} deltaX
+   * @private
+   */
+  drag_(deltaX) {
+    let secondaryTranslate;
+
+    if (deltaX < 0) {
+      this.swipingState_ = SwipingState.SWIPING_TO_NEXT;
+      secondaryTranslate = `translate3d(calc(100% + ${deltaX}px), 0, 0)`;
+    } else {
+      this.swipingState_ = SwipingState.SWIPING_TO_PREVIOUS;
+      secondaryTranslate = `translate3d(calc(${deltaX}px - 100%), 0, 0)`;
+    }
+
+    const story = this.stories_[this.currentIdx_];
+    const iframe = this.iframes_[story[IFRAME_IDX]];
+    const translate = `translate3d(${deltaX}px, 0, 0)`;
+
+    setStyles(iframe, {
```

Comment:
Did you try to wrap these into requestAnimationFrame to see if that improves performance? It's basically what `mutateElement` would do.

Reply:
Done.